### ISSUE TITLE
Line debugger fails on homonyms

### ISSUE THREAD

**Issue by vincentb1** [jdee-emacs/jdee#153]
This is a duplicate of [bug №30 on Sourceforge](https://sourceforge.net/p/jdee/bugs/30/), and I am coming with a fix.

**Comment by vincentb1** [comment]
Attached is a patch file.

[SF_bug30_GH_bug153.diff.zip](https://github.com/jdee-emacs/jdee/files/2285535/SF_bug30_GH_bug153.diff.zip)

This is a starting point, it suffers several remaining issues:

- first it has several spurious changes due to some latin1/utf8 transcoding stuff, I need to make some hook so that when I open an Elisp file in the JDEE work area coding scheme remains UTF-8
- also it has an issue that it tackles two things at a time:
  - it solves the bug
  - but it also brings a code optimisation by using a cache `jdee-source-path-hashmap-cache` hashmap. 
- finally when fixing the bug I realize the the root cause why I fell into the bug is the poor documentation of `jdee-sourcepath` variable, and this patch does not fix this issue.

This code optimisation is incomplete, because I think that this would bring some problems when you have several instances of the same projects in parallel, ie the same full classname in one project would map to a source file in another project. I did not find any simple way to solve this. Probably one way would be to combine the full-classname with a unique project id in the map keys. Another more radical way would be to rewrite the whole function as a method of the debugger object and make the hashmap a member of this object,

OK, it would probably be more reasonable to have a phased approach and fix first the bug w/o the optimisation.

**Comment by vincentb1** [comment]
Concerning the documentation of `jdee-sourcepath`, here is what I mean. To go on with the jPicEdt example which I gave in the [orginal bug submission](https://sourceforge.net/p/jdee/bugs/30/), I have `jdee-sourcepath` containing this two paths: `/jpicedt-project/jpicedy/format/output/tikz` and `/jpicedt-project/jpicedt/format/output/pstricks`, the current JDEE code bogously selects the first path in the list w/o checking that the classname is `jpicedt.format.output.pstricks.PicEllipseFormatter`, so it wrongly sends the line debugger at the source of `jpicedt.format.output.tikz.PicEllipseFormatter` .
This would be solved if the `jdee-sourcepath` just contained `/jpicedt-project` (that would incidentally also make it quite simpler and maintainable, because currently it is a very long list of paths for jPicEdt…). So in a nutshell my problem could be seen as a wrong configuration of JDEE because it could be worked around by configuring JDEE otherwise without any code fix.
However, I believe that is still a problem to be fixed in JDEE, especially the docstring of jdee-sourcepath should be improved — I will provide another patch to propose a fix for that.
